### PR TITLE
Belos,Tpetra: Fix #4626, #4639, #4633

### LINKED ISSUE #4626: Belos::MultiVecTraits: Tpetra specialization is slow on GPU, because it creates local MV & thus does extra CUDA allocations
@trilinos/belos @trilinos/tpetra @cgcgcg 

The Tpetra specialization of `Belos::MultiVecTraits` has two methods that need to create temporary "local" MultiVector instances.  This creation does a new DualView allocation each time.  This makes running on the GPU slow, especially for classical Gram-Schmidt in GMRES, or in general whenever doing `X^T * Y` or `X^H * Y` for either X or Y having multiple columns. 

I am working on a fix.  The idea is to maintain a static DualView "pool" that the local MultiVector can view.  One must be careful not to let any DualView instance persist past `Kokkos::finalize`.

### LINKED ISSUE #4633: Tpetra::MultiVector constructor that takes a device View uses create_mirror instead of create_mirror_view
@trilinos/tpetra @vbrunini 

In most cases, `Tpetra::MultiVector`'s host View is the result of `Kokkos::create_mirror_view` on its device View.  However, in one case:

https://github.com/trilinos/Trilinos/blob/5c9a2ec8019b310b5b74a22e23fcbea11486f781/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp#L427

Tpetra creates the host View from the input device View using `Kokkos::create_mirror`.  This means that, unlike in all other cases, the host View is always a separate allocation (even if not building with CUDA!) from the device View.  That's weird, and it may be causing issues relating to my fixes for #4626.

Note the comment above that line of code:
```
    // The difference between create_mirror and create_mirror_view, is
    // that the latter copies to host.  We don't necessarily want to
    // do that; we just want to allocate the memory.
```
`git blame` dates that to 2015.  It could have been the case back then that `create_mirror_view` copied to host.  It certainly does not now.

### LINKED ISSUE #4639: Tpetra::MultiVector::reduce broken if getStride() > getLocalLength()
@trilinos/tpetra @trilinos/belos @cgcgcg 

## Current Behavior

If a `Tpetra::MultiVector` has `getStride() > getLocalLength()`, then `reduce()` gives incorrect results.

## Motivation and Context

I discovered this while working on a fix for #4626, a Belos performance issue on GPUs.  My original attempted fix created MultiVectors from DualViews with `stride(1) > extent(0)`.  The issue manifested as some Belos tests failing.  It turns out that no Tpetra tests must have been exercising `reduce()` with MultiVectors with this property.

## Possible Solution

I have a fix ready.

## Steps to Reproduce

1. Create `Kokkos::DualView` `dv_orig` with M + S rows and N columns, where M, S, and N are positive integers.
2. `auto dv = Kokkos::subview (dv_orig, std::pair<size_t, size_t> (0, M), Kokkos::ALL ());`
3. Create `Tpetra::MultiVector` with a locally replicated Map (M rows per process, over `MPI_COMM_WORLD`) and `dv`.
4. Call `reduce()` on the MultiVector.  The results are wrong, even in a non-CUDA build.

## Related Issues

* Related to #4626, #4633

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/belos @trilinos/tpetra @cgcgcg @vbrunini 

## Description

This PR supersedes #4640.

  - Fix #4639 (an actual correctness bug in `Tpetra::MultiVector::reduce` that hindered solving #4626)
  - Fix #4633 (a strange thing in Tpetra::MultiVector that could lead to bugs, especially with FEMultiVector)
  - Fix #4626 (a Belos performance issue, especially on GPUs)

## Related Issues

* Closes #4626, #4639, #4633

### CODE REVIEW

**Comment by mhoemmen** [comment]
There's something really quite weird about the test results, especially with the Intel build.  CUDA build passes, though!  I'll mark this as WIP, but please feel free to review.

**Comment by mhoemmen** [comment]
I fixed the build warnings and added an extra check for what I suspect is an Intel compiler bug.  If the tests still fail in the Intel build, then I may need to change DualViewPool's implementation to use a common CUDA UVM memory pool for MultiVectors of all types (so that initialization of the memory pool happens in a nontemplated function), rather than a memory pool per (Scalar, Device) combination.

**Comment by mhoemmen** [comment]
@trilinos/framework PR testing didn't get my new commits.

File: packages/tpetra/core/src/Tpetra_MultiVector_def.hpp [hunk 1]
```diff
@@ -3677,77 +3692,71 @@ namespace Tpetra {
...
-          std::logic_error, ": srcColView and dstColView_dev have different "
-          "dimensions.  Please report this bug to the Tpetra developers.");
-        Kokkos::deep_copy (dstColView, srcColView_dev);
+	Kokkos::deep_copy (A_view, srcView_dev);
+      }
+    }
+    else {
+      for (size_t j = 0; j < numCols; ++j) {
+	const size_t srcCol = this->whichVectors_[j];
+	auto dstColView = Kokkos::subview (A_view, rowRange, j);
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_MultiVector_def.hpp]: @trilinos/framework @GeoffDanielson @william76 @bartlettroscoe @jwillenbring @prwolfe  You can literally see that I've removed the line of code that shows up in PR testing as an unused variable warning.

**Comment by mhoemmen** [comment]
@bartlettroscoe @jwillenbring @william76 @GeoffDanielson @prwolfe @rrdrake @theguruat12 @trilinos/framework @vbrunini @jclause 

PR testing fails to pick up new commits, mysteriously runs out of memory, etc.

**Comment by mhoemmen** [comment]
@kddevin @trilinos/tpetra It's pretty hard to refactor Tpetra and make the deadline for backwards compatibility changes, if we can't get PRs through.

**Comment by mhoemmen** [comment]
Here is the commit in question that fixes the warning, that nevertheless still shows up in PR tests: https://github.com/trilinos/Trilinos/pull/4648/commits/83ad1c2fa5920697d0692fde2c2127fcbe0a5145

**Comment by prwolfe** [comment]
Come see me when you get in – this is worse than you realized.

-----------------------------------------------------------------------
Recommending policy and creating automation to
enable DevOps in the Trilinos and CTH organizations.

**Comment by prwolfe** [comment]
It looks like all PR builds that test stratimikos have been failing since Mar 19, 2019 18:53 UTC (10:53 MDT) This includes one I have in that did not touch any source at all. #4659

**Comment by bartlettroscoe** [comment]
@prwolfe said:

> It looks like all PR builds that test stratimikos have been failing since Mar 19, 2019 18:53 UTC (10:53 MDT) This includes one I have in that did not touch any source at all. #4659

The post-push CI build clearly showed this last night as described in the below email I just sent out.

Are you and the other @trilinos/framework team members getting those CDash emails?  It is important to get those and react quickly to those because that means that eveyrone's PR builds will fail and no one will be able to merge anything (that flows through those packages).

----


From: Bartlett, Roscoe A 
Sent: Wednesday, March 20, 2019 9:30 AM
To: 'trilinos-developers@trilinos.org' <trilinos-developers@trilinos.org>
Subject: Trilinos 'develop' broken?

Hello Trilinos Developers,

FYI: The post-push CI build shown at:

https://testing-vm.sandia.gov/cdash-dev-view/index.php?project=Trilinos&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=61&value1=Linux-GCC-4.8.4-MPI_RELEASE_DEBUG_SHARED_PT_OPENMP_CI&field2=groupname&compare2=61&value2=Continuous&field3=buildstarttime&compare3=83&value3=4%20weeks%20ago

shows that somehow two Stratimikos tests are now failing in the Trilinos ‘develop’ branch in the GCC 4.8.4 OpenMP build (which should match the Trilinos GCC 4.8.4 OpenMP PR build). I think that means that almost all of the Trilinos PRs that change code upstream from Stratimikos will also be broken.

One of the commits from one of the PRs merged to ‘develop’ shown at:

   https://testing-vm.sandia.gov/cdash-dev-view/viewNotes.php?buildid=4746581#!#note0 

must have triggered this. 

Someone should do a root-cause analysis to see how this could have occurred.

Cheers,

-Ross

**Comment by cgcgcg** [comment]
Hm, my PR for the Belos Stratimikos adapter was merged yesterday: #4606

**Comment by prwolfe** [comment]
I get them, but just triaging my email could easily be a full time job. I frequently see issues in other ways or in this case from another e-mail first.

-----------------------------------------------------------------------
Recommending policy and creating automation to
enable DevOps in the Trilinos and CTH organizations.

**Comment by cgcgcg** [comment]
I believe I found the culprit PR #4650 where we changed the some defaults for PseudoBlockGMRES.
This does not explain, however, how this managed to get past the autotester..

**Comment by cgcgcg** [comment]
My attempt at a fix: PR #4677

**Comment by jhux2** [comment]
___
*[packages/tpetra/core/src/Tpetra_MultiVector_def.hpp, line 1815 at r2](https://reviewable.io/reviews/trilinos/trilinos/4648#-LaRuyuY2uC4O6YTlyk_:-LaRuyuY2uC4O6YTlyka:bfu5wjw) ([raw file](https://github.com/trilinos/trilinos/blob/6e83ef970a7e2433fd9815624eb382bd098b8a78/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp#L1815)):*
> ```cpp
>       auto map = x.getMap ();
>       Teuchos::RCP<const Teuchos::Comm<int> > comm =
> 	map.is_null () ? Teuchos::null : map->getComm ();
> ```

Minor quibble -- is this a tab?  reviewable shows a number of red ">>" marks, which I think mean tabs.

**Comment by mhoemmen** [comment]
> Minor quibble -- is this a tab? reviewable shows a number of red ">>" marks, which I think mean tabs.

Oof, could be.  I will review my `.emacs` files to make sure they autoconvert tabs to spaces.

**Comment by mhoemmen** [comment]
I tested with Intel 17.0.1, with OpenMP both on and off, and could not replicate the test failures seen earlier.  Thus, I'm just marking this as RETEST and hoping that the previous fixes will make it all work.

**Comment by mhoemmen** [comment]
@trilinos/tpetra @kddevin I suspect Intel has a bug in its handling of `static` initialization, since the assert fails only in the Intel build.  (It even passes the CUDA build, and it certainly passes my Clang build.)  Regardless, I will need to work around the Intel bug, so I'm marking this PR as WIP.
```
p=0: *** Caught standard std::exception of type 'std::logic_error' :
 
  /scratch/trilinos/workspace/trilinos-folder/Trilinos_pullrequest_intel_17.0.1/Trilinos/packages/belos/tpetra/src/BelosMultiVecTraits_Tpetra.hpp:157:
 
  Throw number = 1
  
  Throw test that evaluated to true: !(d_view_ptr_ != nullptr)
 ```

**Comment by mhoemmen** [comment]
@trilinos/tpetra @kddevin 

I'll roll the dice once more on this one.  If the Intel compiler persists in being buggy, I have a work-around in progress.

**Comment by mhoemmen** [comment]
@trilinos/tpetra The same `!(d_view_ptr_ != nullptr)` nonsense that's clearly impossible.  I have a rewrite in progress that should work around this Intel compiler bug.

**Comment by mhoemmen** [comment]
PR #4747 and PR #4734 (plus follow-ons to the latter) have superseded this PR.  See explanation in #4747.

**Comment by mhoemmen** [comment]
See also branch https://github.com/mhoemmen/Trilinos/tree/TpetraStaticView3 , which contains the aforementioned salvaged and updated code that fixes #4626.

**Comment by mhoemmen** [comment]
Update: See branch https://github.com/mhoemmen/Trilinos/tree/TpetraStaticView4.

**Comment by mhoemmen** [comment]
PR #4765 supersedes this PR.  Thanks all!

**Pull request closed without merging**